src/data/finsum.lean

```diff
@@ -0,0 +1,657 @@
+  ∑ᶠ i : α, f i = finset.sum hf.to_finset f :=
+by rw [finsum_def, dif_pos hf]
+
+lemma finsum_eq_finset_sum' [fintype α] (f : α → M) :
+  ∑ᶠ i : α, f i = (finset.univ' α).sum f :=
+begin
+  rw [finsum_def, dif_pos (set.finite.of_fintype (_ : set α))],
+  refine finset.sum_subset (λ _ _, finset.mem_univ' _) (λ _ _ hx, _),
+    rw set.finite.mem_to_finset at hx,
+    exact function.nmem_support.1 (λ h, hx h),
+end
+
+lemma finsum_in_eq_finset_sum (f : α → M) {s : set α}
+  (hf : (s ∩ function.support f).finite) :
+  ∑ᶠ i in s, f i = ∑ i in hf.to_finset, f i :=
+begin
+  rw [finsum_in_def, finsum_eq_finset_sum],
+  { apply finset.sum_congr,
+    { ext,
+      rw [set.finite.mem_to_finset, set.finite.mem_to_finset, set.mem_inter_iff,
+          function.mem_support, function.mem_support],
+      split,
+      { split_ifs,
+        { exact and.intro h },
+        { intro h2, exact false.elim (h2 rfl) } },
+      { rintro ⟨h1, h2⟩, rwa if_pos h1 } },
+    { intros x hx,
+      rw set.finite.mem_to_finset at hx,
+      exact if_pos hx.1 } },
+  { refine set.finite.subset hf (λ x hx, set.mem_inter _ _),
+    { rw function.mem_support at hx,
+      exact set.mem_of_indicator_ne_zero hx },
+    { change x ∈ function.support (s.indicator f) at hx,
+      rw set.support_indicator at hx,
+      exact hx.2 } },
+end
+
+lemma finsum_in_eq_finset_sum'' (f : α → M) (s : set α)
+  (hf : (function.support f).finite) :
+  ∑ᶠ i in s, f i = (finset.filter (∈ s) hf.to_finset).sum f :=
+begin
+  rw finsum_in_eq_finset_sum f (set.finite.subset hf (set.inter_subset_right _ _)),
+  congr, ext, finish
+end
+
+lemma finsum_in_eq_finset_sum' [fintype α] (f : α → M) (s : set α):
+  ∑ᶠ i in s, f i = ∑ i in s.to_finset, f i :=
+begin
+  rw finsum_in_eq_finset_sum'' f s (set.finite.of_fintype _),
+  conv_rhs { rw ← finset.sum_filter_ne_zero },
+  exact finset.sum_congr (by ext; finish) (λ _ _, rfl),
+end
+
+lemma finsum_in_eq_finset_sum''' (f : α → M) {s : set α} (hs : s.finite) :
+  ∑ᶠ i in s, f i = ∑ i in hs.to_finset, f i :=
+begin
+  rw finsum_in_eq_finset_sum f (set.finite.subset hs (set.inter_subset_left _ _)),
+  conv_rhs { rw ← finset.sum_filter_ne_zero },
+  refine finset.sum_congr _ (λ _ _, rfl),
+  ext, rw [set.finite.to_finset, set.finite.to_finset, finset.mem_filter, set.mem_to_finset,
+           set.mem_to_finset, set.mem_inter_iff, function.mem_support]
+end
+
+lemma finsum_in_eq_finset_sum'''' (f : α → M) (s : finset α) :
+  ∑ᶠ i in ↑s, f i = ∑ i in s, f i :=
+begin
+  rw [finsum_in_eq_finset_sum''' f (finset.finite_to_set s), set.finite.to_finset],
+  congr, ext, simp
+end
+
+lemma finsum_in_eq_zero_of_infinite {f : α → M} {s : set α}
+  (hs : ¬ (s ∩ function.support f).finite) : ∑ᶠ i in s, f i = 0 :=
+begin
+  rw [finsum_in_def, finsum_def, dif_neg],
+  intro h,
+  apply hs,
+  convert h,
+  exact set.support_indicator.symm,
+end
+
+lemma finsum_in_inter_support (f : α → M) (s : set α) :
+  ∑ᶠ i in s, f i = ∑ᶠ i in (s ∩ function.support f), f i :=
```

Comment:
This should be generalized to `∑ᶠ i in s, f i = ∑ᶠ i in t, f i` if `s ∩ function.support f = t ∩ function.support f`, with the hypothesis possibly also written as `\all x \in function.support f, x \in s <-> x \in t`. (This theorem can be either a consequence of that version or vice versa.)

Reply:
done